sys/net/transport_layer/ng_udp/ng_udp.c

```diff
@@ -0,0 +1,256 @@
+
+    /* and forward packet to the network layer */
+    sendto = ng_netreg_lookup(pkt->type, NG_NETREG_DEMUX_CTX_ALL);
+    /* throw away packet if no one is interested */
+    if (sendto == NULL) {
+        DEBUG("udp: cannot send packet because network layer not found\n");
+        ng_pktbuf_release(pkt);
+        return;
+    }
+    /* send packet to network layer */
+    ng_pktbuf_hold(pkt, ng_netreg_num(pkt->type, NG_NETREG_DEMUX_CTX_ALL) - 1);
+    while (sendto != NULL) {
+        ng_netapi_send(sendto->pid, pkt);
+        sendto = ng_netreg_getnext(sendto);
+    }
+}
+
+static void *_event_loop(void *arg)
+{
+    (void)arg;
+    msg_t msg, reply;
+    msg_t msg_queue[NG_UDP_MSG_QUEUE_SIZE];
+    ng_netreg_entry_t netreg;
+
+    /* preset reply message */
+    reply.type = NG_NETAPI_MSG_TYPE_ACK;
+    reply.content.value = (uint32_t)-ENOTSUP;
+    /* initialize message queue */
+    msg_init_queue(msg_queue, NG_UDP_MSG_QUEUE_SIZE);
+    /* register UPD at netreg */
+    netreg.demux_ctx = NG_NETREG_DEMUX_CTX_ALL;
+    netreg.pid = thread_getpid();
+    ng_netreg_register(NG_NETTYPE_UDP, &netreg);
+
+    /* dispatch NETAPI messages */
+    while (1) {
+        msg_receive(&msg);
+        switch (msg.type) {
+            case NG_NETAPI_MSG_TYPE_RCV:
+                DEBUG("udp: NG_NETAPI_MSG_TYPE_RCV\n");
+                _receive((ng_pktsnip_t *)msg.content.ptr);
+                break;
+            case NG_NETAPI_MSG_TYPE_SND:
+                DEBUG("udp: NG_NETAPI_MSG_TYPE_SND\n");
+                _send((ng_pktsnip_t *)msg.content.ptr);
+                break;
+            case NG_NETAPI_MSG_TYPE_SET:
+            case NG_NETAPI_MSG_TYPE_GET:
+                msg_reply(&msg, &reply);
+                break;
+            default:
+                DEBUG("udp: received unidentified message\n");
+                break;
+        }
+    }
+
+    /* never reached */
+    return NULL;
+}
+
+int ng_udp_calc_csum(ng_pktsnip_t *hdr, ng_pktsnip_t *pseudo_hdr)
+{
+    uint16_t csum;
+
+    if (hdr == NULL || hdr->next == NULL) {
+        return -EFAULT;
+    }
+    if (hdr->type != NG_NETTYPE_UDP) {
+        return -EBADMSG;
+    }
+
+    csum = _calc_csum(hdr, pseudo_hdr, hdr->next);
+    if (csum == 0) {
+        return -ENOENT;
+    }
+    ((ng_udp_hdr_t *)hdr->data)->checksum = byteorder_htons(csum);
+    return 0;
+}
+
+ng_pktsnip_t *ng_udp_hdr_build(ng_pktsnip_t *payload,
+                               uint8_t *src, size_t src_len,
+                               uint8_t *dst, size_t dest_len) {
+    ng_pktsnip_t *res;
+    ng_udp_hdr_t *hdr;
+
+    /* allocate header */
+    res = ng_pktbuf_add(payload, NULL, sizeof(ng_udp_hdr_t), NG_NETTYPE_UDP);
+    if (res == NULL) {
+        return NULL;
+    }
+    /* initialize header */
+    hdr = (ng_udp_hdr_t *)res->data;
+    hdr->src_port = byteorder_htons(*((uint16_t *)src));
+    hdr->dst_port = byteorder_htons(*((uint16_t *)dst));
```

Comment:
Yes, everything except in-header data must be in host byte order. Otherwise uncontrollable, hard to track down back-and-forth conversions like in the old stack happen.